sycl/plugins/unified_runtime/ur/adapters/level_zero/device.cpp

```diff
@@ -1049,9 +1049,12 @@ ur_result_t urDeviceRetain(ur_device_handle_t Device) {
 
 ur_result_t urDeviceRelease(ur_device_handle_t Device) {
   // Root devices are destroyed during the piTearDown process.
-  if (Device->isSubDevice()) {
-    if (Device->RefCount.decrementAndTest()) {
-      delete Device;
+  if (Device) {
+    if (Device->isSubDevice()) {
+      if (Device->RefCount.decrementAndTest()) {
+        delete Device;
+        Device = nullptr;
```

Comment:
This change looks like a workaround specific to L-adapter and to release of the device only.
If we cannot control the order of unloading SYCL/UR/adapters then we should look into a more robust solution we did in https://github.com/intel/llvm/pull/8948:  adapters/UR would return *_RESULT_ERROR_UNINITIALIZED after release of their resources, and avoid further use of those resources from the higher level.

Reply:
agree with @smaslov-intel . This looks like a WA. Moreover, from UR spec, urDeviceRelease would fail with validation layer if NULL == hDevice, so that means that we should have never gotten to this point if device is null.

@nrspruit : could we try to see if we can do as suggested by @smaslov-intel above?